pulumi/infra/ec2_cluster.py

```diff
@@ -0,0 +1,92 @@
+from typing import Dict, List, Optional, Sequence
+
+import pulumi_aws as aws
+
+import pulumi
+from pulumi import Output
+
+
+class Ec2Cluster(pulumi.ComponentResource):
+    def __init__(
+        self,
+        name: str,
+        # For quorum purposes, you probably want something >=3 and odd.
+        num_instances: int,
+        ami: str,
+        instance_type: str,
+        iam_instance_profile: aws.iam.InstanceProfile,
+        vpc_private_subnet: aws.ec2.Subnet,
+        vpc_security_group_ids: Sequence[Output[str]],
+        instance_tags: Dict[str, str],
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ) -> None:
+        super().__init__("grapl:Ec2ClusterResource", name=name, props=None, opts=opts)
+        child_opts = pulumi.ResourceOptions(parent=self)
+
+        self.instances = self._build_instances(
+            name=f"{name}-cluster",
+            ami=ami,
+            instance_type=instance_type,
+            num_instances=num_instances,
+            iam_instance_profile=iam_instance_profile,
+            vpc_private_subnet=vpc_private_subnet,
+            vpc_security_group_ids=vpc_security_group_ids,
+            instance_tags=instance_tags,
+            child_opts=child_opts,
+        )
+        self.register_outputs(
+            {
+                "instances": self.instances,
+            }
+        )
+
+    @staticmethod
+    def _build_instances(
+        name: str,
+        ami: str,
+        instance_type: str,
+        num_instances: int,
+        iam_instance_profile: aws.iam.InstanceProfile,
+        vpc_private_subnet: aws.ec2.Subnet,
+        vpc_security_group_ids: Sequence[Output[str]],
+        instance_tags: Dict[str, str],
+        child_opts: pulumi.ResourceOptions,
+    ) -> List[aws.ec2.Instance]:
+        instances = []
+        for i in range(0, num_instances):
+            network_interface = aws.ec2.NetworkInterface(
+                f"ec2-eni-{name}-{i}",
+                subnet_id=vpc_private_subnet.id,
+                security_groups=vpc_security_group_ids,
```

Comment:
Okay, so in one case we just pass in the `id` and in the other case we pass in the real object.

I actually like the real-object more (just form a typing perspective)